### PR TITLE
Replay Verify scheduler

### PR DESCRIPTION
**Pull request by areshand**
## Description

1. A basic scheduler that can schedule tasks based on config and collect results for reporting.
Next step is to put it on a real cluster and add it to github workflow
2. automation for creating snapshot from archive node, creating disk, PV, PVC for k8s cluster
2. fix a bug on replay-on-archive when no txn to be executed

## How Has This Been Tested?

Tested on my local kube cluster

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Move Compiler
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: testsuite/replay-verify/main.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        sorted_skips = sorted(self.ranges_to_skip, key=lambda x: x[0])
+
+        while current < self.end_version:
+            # Check if current position falls within any skip range
+            should_skip = False
+            for skip_start, skip_end in sorted_skips:
+                if skip_start <= current < skip_end:
+                    current = skip_end  # Jump to end of skip range
+                    should_skip = True
+                    break
```
Comment (msmouse) [inline testsuite/replay-verify/main.py:219]: nit: should pop things out of the list as you create ranges? it probably doesn't matter much but feels like a meaningful waste.

File: testsuite/replay-verify/main.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+                    or self.current_workers[i].is_completed()
+                ):
+                    if (
+                        self.current_workers[i] is not None
+                        and self.current_workers[i].is_completed()
+                    ):
+                        self.collect_logs_from_completed_pods(self.current_workers[i])
+                    if len(self.tasks) == 0:
+                        break
+                    chunk = self.tasks.pop(0)
```
Comment (msmouse) [inline testsuite/replay-verify/main.py:246]: nit: you are using three concepts "range", "task" and "chunk" to refer to exactly the same thing :trollface:

File: testsuite/replay-verify/main.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+                        chunk[0],
+                        chunk[1],
+                        self.get_label(),
+                        self.network,
+                        self.namespace,
+                    )
+                    self.current_workers[i] = worker_pod
+                    worker_pod.start()
+                if self.current_workers[i] is not None:
+                    print(f"Checking worker {self.current_workers[i].get_phase()}")
```
Comment (msmouse) [inline testsuite/replay-verify/main.py:257]: Do you want to print out `i` maybe?

File: testsuite/replay-verify/main.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        print("All tasks have been scheduled")
+
+    def collect_logs_from_completed_pods(self, worker_pod):
+        if worker_pod.has_txn_mismatch():
+            logger.info(
+                f"Worker {self.current_workers[i].name} failed with txn mismatch"
+            )
+            self.txn_mismatch_logs.append(worker_pod.get_humio_log_link())
+
+        if worker_pod.is_failed():
```
Comment (msmouse) [inline testsuite/replay-verify/main.py:268]: Shall we log something under this as well?

**Review by msmouse — APPROVED** [review]
🪄

File: testsuite/replay-verify/README.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
+context: https://www.notion.so/aptoslabs/WIP-Replay-Verify-Service-Design-1128b846eb728029bc51fcafb3de6fca?pvs=4
```
Comment (perryjrandall) [inline testsuite/replay-verify/README.md:1]: I dont think this is externally visible? Can you include the tl;dr here or in a publicly accessible file?

File: testsuite/replay-verify/replay-verify-worker-template.yaml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+kind: Pod  
+metadata:  
+  name: worker-pod
+  labels:
+    run: some-label
+spec:  
+  serviceAccountName: default  
+  restartPolicy: Never  # Pod restarts only if it fails  
+  containers:  
+    - name: replay-verify-worker
```
Comment (perryjrandall) [inline testsuite/replay-verify/replay-verify-worker-template.yaml:11]: probably gonna wanna generate this in code

**Review by perryjrandall — COMMENTED** [review]
You might wanna add replay-verify as a tool in poetry so you can poetry run the tool

https://gist.github.com/perryjrandall/cc38e1a47c49d7da807a512ce7283796

File: testsuite/replay-verify/main.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
+import yaml
+from kubernetes import client, config
+from kubernetes.client.rest import ApiException
+import time
+import logging
+import os
+from enum import Enum
+
+
+# Hyperparameters
```
[the remark this answers is not in the document]
Comment (areshand) [reply testsuite/replay-verify/main.py:10]: removed the ones that are supposed to passed by config
The rest should be fixed

File: testsuite/replay-verify/main.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+        container["env"] = []
+
+    # Check if the environment variable already exists
+    for env_var in container["env"]:
+        if env_var["name"] == name:
+            env_var["value"] = value
+            return
+
+    # If it doesn't exist, add it
+    container["env"].append({"name": name, "value": value})
```
[the remark this answers is not in the document]
Comment (areshand) [reply testsuite/replay-verify/main.py:43]: this is a vector of map/dict and each dict is {"name": name, "value": value}

File: testsuite/replay-verify/replay-verify-worker-template.yaml (hunk 6 above)
Comment (areshand) [reply testsuite/replay-verify/replay-verify-worker-template.yaml:11]: this is updated dynamically in` start() `function. This file serves as a template

**Pull request merged**